### ISSUE TITLE
Add an "explanation for a 4 years old" of what the project is about

### ISSUE THREAD

**Issue by ReblochonMasque** [meson800/tame#4]
Hello @meson800 ,

I was a bit confused after reading through the `readme`, so I had a look at the code, and I am now even more confused.O_o
`tame` is an interesting package, but it is hard to understand how to use it; is it even usable at this time? I am not sure.

Thanks for clarifying a little bit.
Cheers

**Comment by meson800** [comment]
Thanks for the interest! I've gotten somewhat sidetracked in my own research, and had some classes to take last term. Next term I also have more exams next term, so likely won't be super free to get this towards a version 1 in the next few months. I welcome help though!

 As to the roadmap, it's not currently usable because there isn't actually a fully working CLI or GUI. A lot of the backend is written, so most of the functionality can be accessed via Python directly. There were some unexpected issues to solve (specifically, what if someone specifies a circular dependency? How do we handle that properly, or do we restrict people to non-circular methods? I chose the first, which involves a decent amount of graph algorithms.

Remaining tasks are then:
1. Check that inheritance works properly (e.g. you inherit your parents, such that it is searchable, so that you can do searches like 'which microscopy experiments involve this gene construct?').
2. Finish up the backend tests to make sure we don't trigger regressions.
3. Write a CLI interface.
4. Write read-the-docs style documentation so it's usable.
5. Write a GUI interface (currently I'm imagining something Zotero-like).

**Comment by ReblochonMasque** [comment]
Hey, thanks for the quick reply.  
I was looking at master, but the action is in the Develop branch; that's a bit more helpful.  

I think `tame` might intend to do what I am looking for, but it is hard to tell: the project is in desperate need of a 20 lines explanation of what it intends to do precisely; right now, it assumes that the reader knows what good and bad habits biologists have; however, having no familiarity with that domain, I am still scratching my head after spending 30 minutes reading everything I could find, including the code in the 'not-so-obvious-to-find' branch.

For instance, the **`Rationale`** paragraph in the `readme`reads more like a thesis abstract than a 'user friendly': "hey welcome, this is what we do ..., we solve this problem ..., you will find it useful if you need to ...", or if you've ever needed to ...

I understand you are busy, but if you could find the time to push an 'explanation for a 4 years old', that would be of tremendous help to me, and to others too.

Cheers

**Comment by meson800** [comment]
To be fair, the `develop` branch stuff is a well known git layout called [git flow](https://nvie.com/posts/a-successful-git-branching-model/) that's been around for a while, but definitely point taken; I will add this to the README on master for those not familiar :)

I'll probably write up something within the next week. The current README is pretty much as you say, it was notes that I wrote up for myself for implementation sake, it's definitely not what I imagine the final documentation to be like.

For a short preview:

The basic question `tame` tries to answer is how to we add description and context to files we generate? Images come with embedded EXIF metadata that stores the camera settings, lens and sensor size, date, and even sometimes GPS location. Certain file formats like HDF5 allow you to create "self-descriptive" datasets, where everything you need to interpret the data is given within the file. This is not true for most files.

In the more biology realm, this could mean microscopy data, single-cell data, or sequencing data, where the full context (such as what organism did we use? Did we do any engineering modifications to it? What experimental conditions did we use?) is not included in the files we generate.

For more of a software engineering/sysadmin perspective, maybe we are making a log aggregator, which stores a bunch of log files from different servers. There's some obvious metadata that we might want to attach (which server a log file came from), but there is also less obvious stuff that would make sense to be inherited, like specific software configurations that servers might share. `tame` would let you do stuff like have a metadata file representing a server, which depends on other metadata files that define configurations for certain software. Then, you could do operations like "pull up all of the tracked log files for servers that are running an Apache webserver with version < x"

Tame is then a way of having human-editable metadata (stored in .yaml files). At it's most basic, you can assign arbitrary metadata tags (like name, or date last updated, or random strings, etc) to certain files, but there is enough power underneath to do these complex hierarchies (say logfile blah.log comes from server `alpha`, which itself has a certain python configuration and a different docker configuration). You can do all of this stuff just with a text editor, but tame then gives you the ability to search and operate on this rich metadata information.

That was the quick off-the-cuff version; does that make some more sense, or more basic? Also, can you describe what your usecase is more?

**Comment by ReblochonMasque** [comment]
Wonderful, thank you! I think I better understand now.

My use case is to attach 'informal' (loosely defined?) metadata to the files underlying a text based notes management system; I think `tame` covers my needs; I'll take a closer look.

Cheers

**Comment by meson800** [comment]
If you don't need metadata hierarchy, you might just look into using YAML directly. It excels at doing this type of loosely defined metadata, so most of `tame` is just software tools for the complexities that would come up with manually managing hierarchies of metadata or doing stuff like gathering files with certain metadata tags together.

The best YAML example I've seen is [here](https://learnxinyminutes.com/docs/yaml/), you can use `pyyaml` in Python to read/write YAML (which is one of the things that `tame` is doing).

**Comment by ReblochonMasque** [comment]
Yes, you might be correct, thank you, IDK yet. I am at the stage of gathering information. Would you by chance have links to good reading on metadata hierarchies.
